# Mod 5 Project - Using Machine Learning to Predict Coronavirus

This project uses daily updated coronavirus data from John Hopkins Hospital

### STEP 1: IMPORT ALL NECCESSARY LIBRARIES

In [1]:
# import all neccessary libraries
import warnings
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.pylab import rcParams
import numpy as np
import seaborn as sns

import statsmodels.api as sm
from statsmodels.tools.eval_measures import rmse
from statsmodels.tsa.stattools import adfuller
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.statespace.sarimax import SARIMAX
from statsmodels.graphics.tsaplots import plot_acf,plot_pacf # for determining (p,q) orders
from statsmodels.tsa.seasonal import seasonal_decompose      # for ETS Plots
from pmdarima import auto_arima                              # for determining ARIMA orders

from sklearn.metrics import mean_squared_error
from tqdm import tqdm
import itertools

# ignore all harmless waring to keep the notebook clean
warnings.filterwarnings('ignore')

# keep the plot inline in notebookb
%matplotlib inline

### STEP 2: LOAD DATA
#### We have 3 data sets: Confirmed cases, Death cases & Recovered cases

In [2]:
# load data
url = 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/'
df_confirmed = pd.read_csv(url + 'time_series_19-covid-Confirmed.csv')
df_deaths = pd.read_csv(url + 'time_series_19-covid-Deaths.csv')
df_recovered = pd.read_csv(url + 'time_series_19-covid-Recovered.csv')

### STEP 3: INITIAL DATA ANALYSIS

### A/ INITIAL ANALYSIS - df_confirmed

In [3]:
df_confirmed.head()

Province/State Country/Region      Lat      Long  1/22/20  1/23/20  1/24/20  \
0            NaN       Thailand  15.0000  101.0000        2        3        5   
1            NaN          Japan  36.0000  138.0000        2        1        2   
2            NaN      Singapore   1.2833  103.8333        0        1        3   
3            NaN          Nepal  28.1667   84.2500        0        0        0   
4            NaN       Malaysia   2.5000  112.5000        0        0        0   

   1/25/20  1/26/20  1/27/20  ...  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  \
0        7        8        8  ...      50       53       59       70       75   
1        2        4        4  ...     511      581      639      639      701   
2        3        4        5  ...     150      160      178      178      200   
3        1        1        1  ...       1        1        1        1        1   
4        3        4        4  ...     117      129      149      149      197   

   3/14/20  3/15/20  3/16/20  3/17/20  3/18/20  
0       82      114      147      177      212  
1      773      839      825      878      889  
2      212      226      243      266      313  
3        1        1        1        1        1  
4      238      428      566      673      790  

[5 rows x 61 columns]

In [4]:
df_confirmed.tail()

Province/State  Country/Region      Lat     Long  1/22/20  1/23/20  \
457          Aruba     Netherlands  12.5186 -70.0358        0        0   
458            NaN          Zambia -15.4167  28.2833        0        0   
459            NaN        Djibouti  11.8251  42.5903        0        0   
460            NaN     Gambia, The  13.4432 -15.3101        0        0   
461     Montserrat  United Kingdom  16.7425 -62.1874        0        0   

     1/24/20  1/25/20  1/26/20  1/27/20  ...  3/9/20  3/10/20  3/11/20  \
457        0        0        0        0  ...       0        0        0   
458        0        0        0        0  ...       0        0        0   
459        0        0        0        0  ...       0        0        0   
460        0        0        0        0  ...       0        0        0   
461        0        0        0        0  ...       0        0        0   

     3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  3/18/20  
457        0        2        2        2        2        3        4  
458        0        0        0        0        0        0        2  
459        0        0        0        0        0        0        1  
460        0        0        0        0        0        1        1  
461        0        0        0        0        0        0        1  

[5 rows x 61 columns]

In [5]:
df_confirmed.describe(include=['object', 'bool'])

Province/State Country/Region
count              319            462
unique             317            152
top     Grand Princess             US
freq                 2            247

In [6]:
df_confirmed.describe()

Lat        Long     1/22/20     1/23/20     1/24/20     1/25/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean    31.742141  -36.138645    1.201299    1.413420    2.036797    3.103896   
std     18.815664   80.217674   20.708297   20.792959   25.900208   36.062942   
min    -41.454500 -157.858400    0.000000    0.000000    0.000000    0.000000   
25%     27.611300  -92.891550    0.000000    0.000000    0.000000    0.000000   
50%     37.831650  -74.304550    0.000000    0.000000    0.000000    0.000000   
75%     42.314500   20.989275    0.000000    0.000000    0.000000    0.000000   
max     71.706900  174.886000  444.000000  444.000000  549.000000  761.000000   

           1/26/20      1/27/20      1/28/20      1/29/20  ...        3/9/20  \
count   462.000000   462.000000   462.000000   462.000000  ...    462.000000   
mean      4.584416     6.335498    12.073593    13.346320  ...    245.803030   
std      50.412877    67.934111   166.507245   167.590443  ...   3216.131465   
min       0.000000     0.000000     0.000000     0.000000  ...      0.000000   
25%       0.000000     0.000000     0.000000     0.000000  ...      0.000000   
50%       0.000000     0.000000     0.000000     0.000000  ...      1.000000   
75%       0.000000     0.000000     0.000000     0.000000  ...      6.000000   
max    1058.000000  1423.000000  3554.000000  3554.000000  ...  67743.000000   

            3/10/20       3/11/20       3/12/20       3/13/20       3/14/20  \
count    462.000000    462.000000    462.000000    462.000000    462.000000   
mean     256.692641    272.435065    277.798701    314.270563    337.865801   
std     3228.733859   3254.587449   3262.349350   3334.667826   3397.248691   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      1.000000   
75%        7.000000     10.000000     15.000000     20.500000     27.000000   
max    67760.000000  67773.000000  67781.000000  67786.000000  67790.000000   

            3/15/20       3/16/20       3/17/20       3/18/20  
count    462.000000    462.000000    462.000000    462.000000  
mean     362.437229    392.915584    426.714286    465.173160  
std     3469.641132   3555.364285   3655.218904   3791.091715  
min        0.000000      0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000      0.000000  
50%        1.000000      1.000000      2.000000      3.000000  
75%       33.750000     49.750000     63.500000     75.000000  
max    67794.000000  67798.000000  67799.000000  67800.000000  

[8 rows x 59 columns]

In [7]:
# count the null columns
null_columns = df_confirmed.columns[df_confirmed.isnull().any()]
df_confirmed[null_columns].isnull().sum().head()

Province/State    143
dtype: int64

In [8]:
# glance look at the null data
pd.set_option('display.max_columns', 999)
print(df_confirmed[df_confirmed.isnull().any(axis=1)][null_columns].head())

  Province/State
0            NaN
1            NaN
2            NaN
3            NaN
4            NaN


In [9]:
# glance look at the null data
pd.set_option('display.max_rows', 999)
print(df_confirmed[df_confirmed.isnull().any(axis=1)][null_columns])

    Province/State
0              NaN
1              NaN
2              NaN
3              NaN
4              NaN
9              NaN
10             NaN
11             NaN
12             NaN
13             NaN
14             NaN
15             NaN
16             NaN
17             NaN
18             NaN
20             NaN
21             NaN
23             NaN
24             NaN
25             NaN
26             NaN
27             NaN
28             NaN
29             NaN
30             NaN
31             NaN
32             NaN
33             NaN
34             NaN
35             NaN
36             NaN
37             NaN
38             NaN
39             NaN
40             NaN
41             NaN
42             NaN
43             NaN
44             NaN
45             NaN
46             NaN
47             NaN
48             NaN
50             NaN
51             NaN
52             NaN
53             NaN
54             NaN
55             NaN
56             NaN
57             NaN
58          

In [10]:
len(df_confirmed['Province/State'].unique())

318

In [11]:
print(df_confirmed['Province/State'].unique())

[nan 'British Columbia' 'New South Wales' 'Victoria' 'Queensland'
 'South Australia' 'From Diamond Princess' 'Western Australia' 'Tasmania'
 'Northern Territory' 'Ontario' 'Alberta' 'Quebec' 'Washington' 'New York'
 'California' 'Massachusetts' 'Diamond Princess' 'Grand Princess'
 'Georgia' 'Colorado' 'Florida' 'New Jersey' 'Oregon' 'Texas' 'Illinois'
 'Pennsylvania' 'Iowa' 'Maryland' 'North Carolina' 'South Carolina'
 'Tennessee' 'Virginia' 'Arizona' 'Indiana' 'Kentucky'
 'District of Columbia' 'Nevada' 'New Hampshire' 'Minnesota' 'Nebraska'
 'Ohio' 'Rhode Island' 'Wisconsin' 'Connecticut' 'Hawaii' 'Oklahoma'
 'Utah' 'Kansas' 'Louisiana' 'Missouri' 'Vermont' 'Alaska' 'Arkansas'
 'Delaware' 'Idaho' 'Maine' 'Michigan' 'Mississippi' 'Montana'
 'New Mexico' 'North Dakota' 'South Dakota' 'West Virginia' 'Wyoming'
 'Hubei' 'France' 'Guangdong' 'Henan' 'Zhejiang' 'Hunan' 'Anhui' 'Jiangxi'
 'Shandong' 'Jiangsu' 'Chongqing' 'Sichuan' 'Heilongjiang' 'Denmark'
 'Beijing' 'Shanghai' 'Hebei' 'Fuji

In [12]:
len(df_confirmed['Country/Region'].unique())

152

In [13]:
print(df_confirmed['Country/Region'].unique())

['Thailand' 'Japan' 'Singapore' 'Nepal' 'Malaysia' 'Canada' 'Australia'
 'Cambodia' 'Sri Lanka' 'Germany' 'Finland' 'United Arab Emirates'
 'Philippines' 'India' 'Italy' 'Sweden' 'Spain' 'Belgium' 'Egypt'
 'Lebanon' 'Iraq' 'Oman' 'Afghanistan' 'Bahrain' 'Kuwait' 'Algeria'
 'Croatia' 'Switzerland' 'Austria' 'Israel' 'Pakistan' 'Brazil' 'Georgia'
 'Greece' 'North Macedonia' 'Norway' 'Romania' 'Estonia' 'San Marino'
 'Belarus' 'Iceland' 'Lithuania' 'Mexico' 'New Zealand' 'Nigeria'
 'Ireland' 'Luxembourg' 'Monaco' 'Qatar' 'Ecuador' 'Azerbaijan' 'Armenia'
 'Dominican Republic' 'Indonesia' 'Portugal' 'Andorra' 'Latvia' 'Morocco'
 'Saudi Arabia' 'Senegal' 'Argentina' 'Chile' 'Jordan' 'Ukraine' 'Hungary'
 'Liechtenstein' 'Poland' 'Tunisia' 'Bosnia and Herzegovina' 'Slovenia'
 'South Africa' 'Bhutan' 'Cameroon' 'Colombia' 'Costa Rica' 'Peru'
 'Serbia' 'Slovakia' 'Togo' 'Malta' 'Martinique' 'Bulgaria' 'Maldives'
 'Bangladesh' 'Paraguay' 'Albania' 'Cyprus' 'Brunei' 'US' 'Burkina Faso'
 'Holy See'

### B/ INITIAL ANALYSIS - df_deaths

In [14]:
df_deaths.head()

Province/State Country/Region      Lat      Long  1/22/20  1/23/20  1/24/20  \
0            NaN       Thailand  15.0000  101.0000        0        0        0   
1            NaN          Japan  36.0000  138.0000        0        0        0   
2            NaN      Singapore   1.2833  103.8333        0        0        0   
3            NaN          Nepal  28.1667   84.2500        0        0        0   
4            NaN       Malaysia   2.5000  112.5000        0        0        0   

   1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  \
0        0        0        0        0        0        0        0       0   
1        0        0        0        0        0        0        0       0   
2        0        0        0        0        0        0        0       0   
3        0        0        0        0        0        0        0       0   
4        0        0        0        0        0        0        0       0   

   2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  \
0       0       0       0       0       0       0       0       0        0   
1       0       0       0       0       0       0       0       0        0   
2       0       0       0       0       0       0       0       0        0   
3       0       0       0       0       0       0       0       0        0   
4       0       0       0       0       0       0       0       0        0   

   2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        1        1        1        1        1        1   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  \
0        0        0        0        0        0        0        0        0   
1        1        1        1        1        1        1        1        2   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  \
0        0        0        0       1       1       1       1       1       1   
1        4        4        5       6       6       6       6       6       6   
2        0        0        0       0       0       0       0       0       0   
3        0        0        0       0       0       0       0       0       0   
4        0        0        0       0       0       0       0       0       0   

   3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  \
0       1       1       1        1        1        1        1        1   
1       6       6      10       10       15       16       19       22   
2       0       0       0        0        0        0        0        0   
3       0       0       0        0        0        0        0        0   
4       0       0       0        0        0        0        0        0   

   3/15/20  3/16/20  3/17/20  3/18/20  
0        1        1        1        1  
1       22       27       29       29  
2        0        0        0        0  
3        0        0        0        0  
4        0        0        2        2

In [15]:
df_deaths.tail()

Province/State  Country/Region      Lat     Long  1/22/20  1/23/20  \
457          Aruba     Netherlands  12.5186 -70.0358        0        0   
458            NaN          Zambia -15.4167  28.2833        0        0   
459            NaN        Djibouti  11.8251  42.5903        0        0   
460            NaN     Gambia, The  13.4432 -15.3101        0        0   
461     Montserrat  United Kingdom  16.7425 -62.1874        0        0   

     1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  \
457        0        0        0        0        0        0        0        0   
458        0        0        0        0        0        0        0        0   
459        0        0        0        0        0        0        0        0   
460        0        0        0        0        0        0        0        0   
461        0        0        0        0        0        0        0        0   

     2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  \
457       0       0       0       0       0       0       0       0       0   
458       0       0       0       0       0       0       0       0       0   
459       0       0       0       0       0       0       0       0       0   
460       0       0       0       0       0       0       0       0       0   
461       0       0       0       0       0       0       0       0       0   

     2/10/20  2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  \
457        0        0        0        0        0        0        0        0   
458        0        0        0        0        0        0        0        0   
459        0        0        0        0        0        0        0        0   
460        0        0        0        0        0        0        0        0   
461        0        0        0        0        0        0        0        0   

     2/18/20  2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  \
457        0        0        0        0        0        0        0        0   
458        0        0        0        0        0        0        0        0   
459        0        0        0        0        0        0        0        0   
460        0        0        0        0        0        0        0        0   
461        0        0        0        0        0        0        0        0   

     2/26/20  2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  \
457        0        0        0        0       0       0       0       0   
458        0        0        0        0       0       0       0       0   
459        0        0        0        0       0       0       0       0   
460        0        0        0        0       0       0       0       0   
461        0        0        0        0       0       0       0       0   

     3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  \
457       0       0       0       0       0        0        0        0   
458       0       0       0       0       0        0        0        0   
459       0       0       0       0       0        0        0        0   
460       0       0       0       0       0        0        0        0   
461       0       0       0       0       0        0        0        0   

     3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  3/18/20  
457        0        0        0        0        0        0  
458        0        0        0        0        0        0  
459        0        0        0        0        0        0  
460        0        0        0        0        0        0  
461        0        0        0        0        0        0

In [16]:
df_deaths.describe(include=['object', 'bool'])

Province/State Country/Region
count              319            462
unique             317            152
top     Grand Princess             US
freq                 2            247

In [17]:
df_deaths.describe()

Lat        Long     1/22/20     1/23/20     1/24/20     1/25/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean    31.742141  -36.138645    0.036797    0.038961    0.056277    0.090909   
std     18.815664   80.217674    0.790912    0.792178    1.118312    1.861927   
min    -41.454500 -157.858400    0.000000    0.000000    0.000000    0.000000   
25%     27.611300  -92.891550    0.000000    0.000000    0.000000    0.000000   
50%     37.831650  -74.304550    0.000000    0.000000    0.000000    0.000000   
75%     42.314500   20.989275    0.000000    0.000000    0.000000    0.000000   
max     71.706900  174.886000   17.000000   17.000000   24.000000   40.000000   

          1/26/20     1/27/20     1/28/20     1/29/20     1/30/20     1/31/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean     0.121212    0.177489    0.283550    0.287879    0.370130    0.461039   
std      2.420633    3.537051    5.816025    5.816558    7.537859    9.491496   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max     52.000000   76.000000  125.000000  125.000000  162.000000  204.000000   

           2/1/20      2/2/20      2/3/20      2/4/20      2/5/20      2/6/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean     0.560606    0.783550    0.922078    1.064935    1.220779    1.372294   
std     11.584810   16.283441   19.260808   22.284692   25.541149   28.751307   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max    249.000000  350.000000  414.000000  479.000000  549.000000  618.000000   

           2/7/20      2/8/20      2/9/20     2/10/20      2/11/20  \
count  462.000000  462.000000  462.000000  462.000000   462.000000   
mean     1.556277    1.744589    1.961039    2.192641     2.409091   
std     32.519577   36.288129   40.521902   45.313702    49.687127   
min      0.000000    0.000000    0.000000    0.000000     0.000000   
25%      0.000000    0.000000    0.000000    0.000000     0.000000   
50%      0.000000    0.000000    0.000000    0.000000     0.000000   
75%      0.000000    0.000000    0.000000    0.000000     0.000000   
max    699.000000  780.000000  871.000000  974.000000  1068.000000   

           2/12/20      2/13/20      2/14/20      2/15/20      2/16/20  \
count   462.000000   462.000000   462.000000   462.000000   462.000000   
mean      2.419913     2.967532     3.296537     3.606061     3.831169   
std      49.687233    60.945666    67.784979    74.251731    78.903561   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max    1068.000000  1310.000000  1457.000000  1596.000000  1696.000000   

           2/17/20      2/18/20      2/19/20      2/20/20      2/21/20  \
count   462.000000   462.000000   462.000000   462.000000   462.000000   
mean      4.043290     4.344156     4.593074     4.863636     4.872294   
std      83.230766    89.372290    94.396238    99.745451    99.745201   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%     

In [18]:
# count the null columns
null_columns = df_deaths.columns[df_deaths.isnull().any()]
df_deaths[null_columns].isnull().sum().head()

Province/State    143
dtype: int64

In [19]:
# glance look at the null data
pd.set_option('display.max_columns', 999)
print(df_deaths[df_deaths.isnull().any(axis=1)][null_columns].head())

  Province/State
0            NaN
1            NaN
2            NaN
3            NaN
4            NaN


In [20]:
# glance look at the null data
pd.set_option('display.max_rows', 999)
print(df_deaths[df_deaths.isnull().any(axis=1)][null_columns])

    Province/State
0              NaN
1              NaN
2              NaN
3              NaN
4              NaN
9              NaN
10             NaN
11             NaN
12             NaN
13             NaN
14             NaN
15             NaN
16             NaN
17             NaN
18             NaN
20             NaN
21             NaN
23             NaN
24             NaN
25             NaN
26             NaN
27             NaN
28             NaN
29             NaN
30             NaN
31             NaN
32             NaN
33             NaN
34             NaN
35             NaN
36             NaN
37             NaN
38             NaN
39             NaN
40             NaN
41             NaN
42             NaN
43             NaN
44             NaN
45             NaN
46             NaN
47             NaN
48             NaN
50             NaN
51             NaN
52             NaN
53             NaN
54             NaN
55             NaN
56             NaN
57             NaN
58          

In [21]:
len(df_deaths['Province/State'].unique())

318

In [22]:
print(df_deaths['Province/State'].unique())

[nan 'British Columbia' 'New South Wales' 'Victoria' 'Queensland'
 'South Australia' 'From Diamond Princess' 'Western Australia' 'Tasmania'
 'Northern Territory' 'Ontario' 'Alberta' 'Quebec' 'Washington' 'New York'
 'California' 'Massachusetts' 'Diamond Princess' 'Grand Princess'
 'Georgia' 'Colorado' 'Florida' 'New Jersey' 'Oregon' 'Texas' 'Illinois'
 'Pennsylvania' 'Iowa' 'Maryland' 'North Carolina' 'South Carolina'
 'Tennessee' 'Virginia' 'Arizona' 'Indiana' 'Kentucky'
 'District of Columbia' 'Nevada' 'New Hampshire' 'Minnesota' 'Nebraska'
 'Ohio' 'Rhode Island' 'Wisconsin' 'Connecticut' 'Hawaii' 'Oklahoma'
 'Utah' 'Kansas' 'Louisiana' 'Missouri' 'Vermont' 'Alaska' 'Arkansas'
 'Delaware' 'Idaho' 'Maine' 'Michigan' 'Mississippi' 'Montana'
 'New Mexico' 'North Dakota' 'South Dakota' 'West Virginia' 'Wyoming'
 'Hubei' 'France' 'Guangdong' 'Henan' 'Zhejiang' 'Hunan' 'Anhui' 'Jiangxi'
 'Shandong' 'Jiangsu' 'Chongqing' 'Sichuan' 'Heilongjiang' 'Denmark'
 'Beijing' 'Shanghai' 'Hebei' 'Fuji

In [23]:
len(df_deaths['Country/Region'].unique())

152

In [24]:
print(df_deaths['Country/Region'].unique())

['Thailand' 'Japan' 'Singapore' 'Nepal' 'Malaysia' 'Canada' 'Australia'
 'Cambodia' 'Sri Lanka' 'Germany' 'Finland' 'United Arab Emirates'
 'Philippines' 'India' 'Italy' 'Sweden' 'Spain' 'Belgium' 'Egypt'
 'Lebanon' 'Iraq' 'Oman' 'Afghanistan' 'Bahrain' 'Kuwait' 'Algeria'
 'Croatia' 'Switzerland' 'Austria' 'Israel' 'Pakistan' 'Brazil' 'Georgia'
 'Greece' 'North Macedonia' 'Norway' 'Romania' 'Estonia' 'San Marino'
 'Belarus' 'Iceland' 'Lithuania' 'Mexico' 'New Zealand' 'Nigeria'
 'Ireland' 'Luxembourg' 'Monaco' 'Qatar' 'Ecuador' 'Azerbaijan' 'Armenia'
 'Dominican Republic' 'Indonesia' 'Portugal' 'Andorra' 'Latvia' 'Morocco'
 'Saudi Arabia' 'Senegal' 'Argentina' 'Chile' 'Jordan' 'Ukraine' 'Hungary'
 'Liechtenstein' 'Poland' 'Tunisia' 'Bosnia and Herzegovina' 'Slovenia'
 'South Africa' 'Bhutan' 'Cameroon' 'Colombia' 'Costa Rica' 'Peru'
 'Serbia' 'Slovakia' 'Togo' 'Malta' 'Martinique' 'Bulgaria' 'Maldives'
 'Bangladesh' 'Paraguay' 'Albania' 'Cyprus' 'Brunei' 'US' 'Burkina Faso'
 'Holy See'

### C/ INITIAL ANALYSIS - df_recovered

In [25]:
df_recovered.head()

Province/State Country/Region      Lat      Long  1/22/20  1/23/20  1/24/20  \
0            NaN       Thailand  15.0000  101.0000        0        0        0   
1            NaN          Japan  36.0000  138.0000        0        0        0   
2            NaN      Singapore   1.2833  103.8333        0        0        0   
3            NaN          Nepal  28.1667   84.2500        0        0        0   
4            NaN       Malaysia   2.5000  112.5000        0        0        0   

   1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  \
0        0        2        2        5        5        5        5       5   
1        0        1        1        1        1        1        1       1   
2        0        0        0        0        0        0        0       0   
3        0        0        0        0        0        0        0       0   
4        0        0        0        0        0        0        0       0   

   2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  \
0       5       5       5       5       5       5      10      10       10   
1       1       1       1       1       1       1       1       1        4   
2       0       0       0       0       0       0       2       2        2   
3       0       0       0       0       0       0       0       0        0   
4       0       0       0       0       0       1       1       1        1   

   2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
0       10       10       12       12       12       14       15       15   
1        9        9        9        9       12       12       12       13   
2        9       15       15       17       18       18       24       29   
3        0        1        1        1        1        1        1        1   
4        3        3        3        3        7        7        7       13   

   2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  \
0       15       15       17       17       21       21       22       22   
1       18       18       22       22       22       22       22       22   
2       34       34       37       37       51       51       53       62   
3        1        1        1        1        1        1        1        1   
4       15       15       15       15       15       18       18       18   

   2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  \
0       22       28       28      28      31      31      31      31      31   
1       22       22       32      32      32      43      43      43      46   
2       62       62       72      72      78      78      78      78      78   
3        1        1        1       1       1       1       1       1       1   
4       18       18       18      18      18      22      22      22      22   

   3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  \
0      31      31      31       33       34       34       35       35   
1      76      76      76      101      118      118      118      118   
2      78      78      78       78       96       96       97      105   
3       1       1       1        1        1        1        1        1   
4      23      24      24       24       26       26       26       35   

   3/15/20  3/16/20  3/17/20  3/18/20  
0       35       35       41       42  
1      118      144      144      144  
2      105      109      114      114  
3        1        1        1        1  
4       42       42       49       60

In [26]:
df_recovered.head()

Province/State Country/Region      Lat      Long  1/22/20  1/23/20  1/24/20  \
0            NaN       Thailand  15.0000  101.0000        0        0        0   
1            NaN          Japan  36.0000  138.0000        0        0        0   
2            NaN      Singapore   1.2833  103.8333        0        0        0   
3            NaN          Nepal  28.1667   84.2500        0        0        0   
4            NaN       Malaysia   2.5000  112.5000        0        0        0   

   1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  \
0        0        2        2        5        5        5        5       5   
1        0        1        1        1        1        1        1       1   
2        0        0        0        0        0        0        0       0   
3        0        0        0        0        0        0        0       0   
4        0        0        0        0        0        0        0       0   

   2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  \
0       5       5       5       5       5       5      10      10       10   
1       1       1       1       1       1       1       1       1        4   
2       0       0       0       0       0       0       2       2        2   
3       0       0       0       0       0       0       0       0        0   
4       0       0       0       0       0       1       1       1        1   

   2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
0       10       10       12       12       12       14       15       15   
1        9        9        9        9       12       12       12       13   
2        9       15       15       17       18       18       24       29   
3        0        1        1        1        1        1        1        1   
4        3        3        3        3        7        7        7       13   

   2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  \
0       15       15       17       17       21       21       22       22   
1       18       18       22       22       22       22       22       22   
2       34       34       37       37       51       51       53       62   
3        1        1        1        1        1        1        1        1   
4       15       15       15       15       15       18       18       18   

   2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  \
0       22       28       28      28      31      31      31      31      31   
1       22       22       32      32      32      43      43      43      46   
2       62       62       72      72      78      78      78      78      78   
3        1        1        1       1       1       1       1       1       1   
4       18       18       18      18      18      22      22      22      22   

   3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  \
0      31      31      31       33       34       34       35       35   
1      76      76      76      101      118      118      118      118   
2      78      78      78       78       96       96       97      105   
3       1       1       1        1        1        1        1        1   
4      23      24      24       24       26       26       26       35   

   3/15/20  3/16/20  3/17/20  3/18/20  
0       35       35       41       42  
1      118      144      144      144  
2      105      109      114      114  
3        1        1        1        1  
4       42       42       49       60

In [27]:
df_recovered.tail()

Province/State  Country/Region      Lat     Long  1/22/20  1/23/20  \
457          Aruba     Netherlands  12.5186 -70.0358        0        0   
458            NaN          Zambia -15.4167  28.2833        0        0   
459            NaN        Djibouti  11.8251  42.5903        0        0   
460            NaN     Gambia, The  13.4432 -15.3101        0        0   
461     Montserrat  United Kingdom  16.7425 -62.1874        0        0   

     1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  \
457        0        0        0        0        0        0        0        0   
458        0        0        0        0        0        0        0        0   
459        0        0        0        0        0        0        0        0   
460        0        0        0        0        0        0        0        0   
461        0        0        0        0        0        0        0        0   

     2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  \
457       0       0       0       0       0       0       0       0       0   
458       0       0       0       0       0       0       0       0       0   
459       0       0       0       0       0       0       0       0       0   
460       0       0       0       0       0       0       0       0       0   
461       0       0       0       0       0       0       0       0       0   

     2/10/20  2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  \
457        0        0        0        0        0        0        0        0   
458        0        0        0        0        0        0        0        0   
459        0        0        0        0        0        0        0        0   
460        0        0        0        0        0        0        0        0   
461        0        0        0        0        0        0        0        0   

     2/18/20  2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  \
457        0        0        0        0        0        0        0        0   
458        0        0        0        0        0        0        0        0   
459        0        0        0        0        0        0        0        0   
460        0        0        0        0        0        0        0        0   
461        0        0        0        0        0        0        0        0   

     2/26/20  2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  \
457        0        0        0        0       0       0       0       0   
458        0        0        0        0       0       0       0       0   
459        0        0        0        0       0       0       0       0   
460        0        0        0        0       0       0       0       0   
461        0        0        0        0       0       0       0       0   

     3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  \
457       0       0       0       0       0        0        0        0   
458       0       0       0       0       0        0        0        0   
459       0       0       0       0       0        0        0        0   
460       0       0       0       0       0        0        0        0   
461       0       0       0       0       0        0        0        0   

     3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  3/18/20  
457        0        0        0        0        0        0  
458        0        0        0        0        0        0  
459        0        0        0        0        0        0  
460        0        0        0        0        0        0  
461        0        0        0        0        0        0

In [28]:
df_recovered.describe(include=['object', 'bool'])

Province/State Country/Region
count              319            462
unique             317            152
top     Grand Princess             US
freq                 2            247

In [29]:
df_recovered.describe()

Lat        Long     1/22/20     1/23/20     1/24/20     1/25/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean    31.742141  -36.138645    0.060606    0.064935    0.077922    0.084416   
std     18.815664   80.217674    1.302678    1.305795    1.446962    1.495987   
min    -41.454500 -157.858400    0.000000    0.000000    0.000000    0.000000   
25%     27.611300  -92.891550    0.000000    0.000000    0.000000    0.000000   
50%     37.831650  -74.304550    0.000000    0.000000    0.000000    0.000000   
75%     42.314500   20.989275    0.000000    0.000000    0.000000    0.000000   
max     71.706900  174.886000   28.000000   28.000000   31.000000   32.000000   

          1/26/20     1/27/20     1/28/20     1/29/20     1/30/20     1/31/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean     0.112554    0.132035    0.231602    0.272727    0.309524    0.480519   
std      1.961753    2.112326    3.746935    4.122747    4.241295    6.642943   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max     42.000000   45.000000   80.000000   88.000000   90.000000  141.000000   

           2/1/20      2/2/20      2/3/20      2/4/20      2/5/20      2/6/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean     0.614719    1.021645    1.348485    1.844156    2.432900    3.218615   
std      7.953419   13.884852   18.187794   24.628906   30.034522   38.818475   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max    168.000000  295.000000  386.000000  522.000000  633.000000  817.000000   

            2/7/20       2/8/20       2/9/20      2/10/20      2/11/20  \
count   462.000000   462.000000   462.000000   462.000000   462.000000   
mean      4.352814     5.662338     7.021645     8.541126    10.136364   
std      52.935632    68.447192    85.293671   105.412347   125.108269   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max    1115.000000  1439.000000  1795.000000  2222.000000  2639.000000   

           2/12/20      2/13/20      2/14/20      2/15/20      2/16/20  \
count   462.000000   462.000000   462.000000   462.000000   462.000000   
mean     11.147186    13.625541    17.441558    20.335498    23.517316   
std     128.337763   164.333771   225.291481   265.013072   312.417178   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max    2686.000000  3459.000000  4774.000000  5623.000000  6639.000000   

           2/17/20      2/18/20       2/19/20       2/20/20       2/21/20  \
count   462.000000   462.000000    462.000000    462.000000    462.000000   
mean     27.235931    31.064935     34.893939     39.344156     40.887446   
std     369.499941   428.392891    484.897404    552.639567    557.811808   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.000000      0.000000      0.000000      0.000000   
50%       0.000000     0.000

In [30]:
# count the null columns
null_columns = df_recovered.columns[df_recovered.isnull().any()]
df_recovered[null_columns].isnull().sum().head()

Province/State    143
dtype: int64

In [31]:
# glance look at the null data
pd.set_option('display.max_columns', 999)
print(df_recovered[df_recovered.isnull().any(axis=1)][null_columns].head())

  Province/State
0            NaN
1            NaN
2            NaN
3            NaN
4            NaN


In [32]:
# glance look at the null data
pd.set_option('display.max_rows', 999)
print(df_recovered[df_recovered.isnull().any(axis=1)][null_columns])

    Province/State
0              NaN
1              NaN
2              NaN
3              NaN
4              NaN
9              NaN
10             NaN
11             NaN
12             NaN
13             NaN
14             NaN
15             NaN
16             NaN
17             NaN
18             NaN
20             NaN
21             NaN
23             NaN
24             NaN
25             NaN
26             NaN
27             NaN
28             NaN
29             NaN
30             NaN
31             NaN
32             NaN
33             NaN
34             NaN
35             NaN
36             NaN
37             NaN
38             NaN
39             NaN
40             NaN
41             NaN
42             NaN
43             NaN
44             NaN
45             NaN
46             NaN
47             NaN
48             NaN
50             NaN
51             NaN
52             NaN
53             NaN
54             NaN
55             NaN
56             NaN
57             NaN
58          

In [33]:
len(df_recovered['Province/State'].unique())

318

In [34]:
print(df_recovered['Province/State'].unique())

[nan 'British Columbia' 'New South Wales' 'Victoria' 'Queensland'
 'South Australia' 'From Diamond Princess' 'Western Australia' 'Tasmania'
 'Northern Territory' 'Ontario' 'Alberta' 'Quebec' 'Washington' 'New York'
 'California' 'Massachusetts' 'Diamond Princess' 'Grand Princess'
 'Georgia' 'Colorado' 'Florida' 'New Jersey' 'Oregon' 'Texas' 'Illinois'
 'Pennsylvania' 'Iowa' 'Maryland' 'North Carolina' 'South Carolina'
 'Tennessee' 'Virginia' 'Arizona' 'Indiana' 'Kentucky'
 'District of Columbia' 'Nevada' 'New Hampshire' 'Minnesota' 'Nebraska'
 'Ohio' 'Rhode Island' 'Wisconsin' 'Connecticut' 'Hawaii' 'Oklahoma'
 'Utah' 'Kansas' 'Louisiana' 'Missouri' 'Vermont' 'Alaska' 'Arkansas'
 'Delaware' 'Idaho' 'Maine' 'Michigan' 'Mississippi' 'Montana'
 'New Mexico' 'North Dakota' 'South Dakota' 'West Virginia' 'Wyoming'
 'Hubei' 'France' 'Guangdong' 'Henan' 'Zhejiang' 'Hunan' 'Anhui' 'Jiangxi'
 'Shandong' 'Jiangsu' 'Chongqing' 'Sichuan' 'Heilongjiang' 'Denmark'
 'Beijing' 'Shanghai' 'Hebei' 'Fuji

In [35]:
len(df_recovered['Country/Region'].unique())

152

In [36]:
print(df_recovered['Country/Region'].unique())

['Thailand' 'Japan' 'Singapore' 'Nepal' 'Malaysia' 'Canada' 'Australia'
 'Cambodia' 'Sri Lanka' 'Germany' 'Finland' 'United Arab Emirates'
 'Philippines' 'India' 'Italy' 'Sweden' 'Spain' 'Belgium' 'Egypt'
 'Lebanon' 'Iraq' 'Oman' 'Afghanistan' 'Bahrain' 'Kuwait' 'Algeria'
 'Croatia' 'Switzerland' 'Austria' 'Israel' 'Pakistan' 'Brazil' 'Georgia'
 'Greece' 'North Macedonia' 'Norway' 'Romania' 'Estonia' 'San Marino'
 'Belarus' 'Iceland' 'Lithuania' 'Mexico' 'New Zealand' 'Nigeria'
 'Ireland' 'Luxembourg' 'Monaco' 'Qatar' 'Ecuador' 'Azerbaijan' 'Armenia'
 'Dominican Republic' 'Indonesia' 'Portugal' 'Andorra' 'Latvia' 'Morocco'
 'Saudi Arabia' 'Senegal' 'Argentina' 'Chile' 'Jordan' 'Ukraine' 'Hungary'
 'Liechtenstein' 'Poland' 'Tunisia' 'Bosnia and Herzegovina' 'Slovenia'
 'South Africa' 'Bhutan' 'Cameroon' 'Colombia' 'Costa Rica' 'Peru'
 'Serbia' 'Slovakia' 'Togo' 'Malta' 'Martinique' 'Bulgaria' 'Maldives'
 'Bangladesh' 'Paraguay' 'Albania' 'Cyprus' 'Brunei' 'US' 'Burkina Faso'
 'Holy See'

### STEP 4: CLEANING DATA, REDUCE UNNECCESSARY & MISSING DATA FEATURES

#### The column 'Province/State' has 143 missing data, and we actually focus on country information not detail to Province/State. We are going reduce the Province/State column and using total cases by country.

In [37]:
df_confirmed.head()

Province/State Country/Region      Lat      Long  1/22/20  1/23/20  1/24/20  \
0            NaN       Thailand  15.0000  101.0000        2        3        5   
1            NaN          Japan  36.0000  138.0000        2        1        2   
2            NaN      Singapore   1.2833  103.8333        0        1        3   
3            NaN          Nepal  28.1667   84.2500        0        0        0   
4            NaN       Malaysia   2.5000  112.5000        0        0        0   

   1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  \
0        7        8        8       14       14       14       19      19   
1        2        4        4        7        7       11       15      20   
2        3        4        5        7        7       10       13      16   
3        1        1        1        1        1        1        1       1   
4        3        4        4        4        7        8        8       8   

   2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  \
0      19      19      25      25      25      25      32      32       32   
1      20      20      22      22      45      25      25      26       26   
2      18      18      24      28      28      30      33      40       45   
3       1       1       1       1       1       1       1       1        1   
4       8       8      10      12      12      12      16      16       18   

   2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
0       33       33       33       33       33       34       35       35   
1       26       28       28       29       43       59       66       74   
2       47       50       58       67       72       75       77       81   
3        1        1        1        1        1        1        1        1   
4       18       18       19       19       22       22       22       22   

   2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  \
0       35       35       35       35       35       35       37       40   
1       84       94      105      122      147      159      170      189   
2       84       84       85       85       89       89       91       93   
3        1        1        1        1        1        1        1        1   
4       22       22       22       22       22       22       22       22   

   2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  \
0       40       41       42      42      43      43      43      47      48   
1      214      228      241     256     274     293     331     360     420   
2       93       93      102     106     108     110     110     117     130   
3        1        1        1       1       1       1       1       1       1   
4       23       23       25      29      29      36      50      50      83   

   3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  \
0      50      50      50       53       59       70       75       82   
1     461     502     511      581      639      639      701      773   
2     138     150     150      160      178      178      200      212   
3       1       1       1        1        1        1        1        1   
4      93      99     117      129      149      149      197      238   

   3/15/20  3/16/20  3/17/20  3/18/20  
0      114      147      177      212  
1      839      825      878      889  
2      226      243      266      313  
3        1        1        1        1  
4      428      566      673      790

In [38]:
# drop columns that we don't need them
# we focus on data by Country/Region only
dropped_df_confirmed = df_confirmed.drop(['Province/State','Lat', 'Long'], axis=1)
dropped_df_confirmed.head()

Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0       Thailand        2        3        5        7        8        8   
1          Japan        2        1        2        2        4        4   
2      Singapore        0        1        3        3        4        5   
3          Nepal        0        0        0        1        1        1   
4       Malaysia        0        0        0        3        4        4   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0       14       14       14       19      19      19      19      25      25   
1        7        7       11       15      20      20      20      22      22   
2        7        7       10       13      16      18      18      24      28   
3        1        1        1        1       1       1       1       1       1   
4        4        7        8        8       8       8       8      10      12   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0      25      25      32      32       32       33       33       33   
1      45      25      25      26       26       26       28       28   
2      28      30      33      40       45       47       50       58   
3       1       1       1       1        1        1        1        1   
4      12      12      16      16       18       18       18       19   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0       33       33       34       35       35       35       35       35   
1       29       43       59       66       74       84       94      105   
2       67       72       75       77       81       84       84       85   
3        1        1        1        1        1        1        1        1   
4       19       22       22       22       22       22       22       22   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0       35       35       35       37       40       40       41       42   
1      122      147      159      170      189      214      228      241   
2       85       89       89       91       93       93       93      102   
3        1        1        1        1        1        1        1        1   
4       22       22       22       22       22       23       23       25   

   3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  \
0      42      43      43      43      47      48      50      50      50   
1     256     274     293     331     360     420     461     502     511   
2     106     108     110     110     117     130     138     150     150   
3       1       1       1       1       1       1       1       1       1   
4      29      29      36      50      50      83      93      99     117   

   3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
0       53       59       70       75       82      114      147      177   
1      581      639      639      701      773      839      825      878   
2      160      178      178      200      212      226      243      266   
3        1        1        1        1        1        1        1        1   
4      129      149      149      197      238      428      566      673   

   3/18/20  
0      212  
1      889  
2      313  
3        1  
4      790

In [39]:
dropped_df_confirmed.describe(include=['object', 'bool'])

Country/Region
count             462
unique            152
top                US
freq              247

In [40]:
dropped_df_confirmed_total = dropped_df_confirmed.groupby(['Country/Region']).sum()
dropped_df_confirmed_total.head()

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
Country/Region                                                              
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   

                     1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  \
Country/Region                                                            
Afghanistan                0        0        0        0       0       0   
Albania                    0        0        0        0       0       0   
Algeria                    0        0        0        0       0       0   
Andorra                    0        0        0        0       0       0   
Antigua and Barbuda        0        0        0        0       0       0   

                     2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  \
Country/Region                                                                
Afghanistan               0       0       0       0       0       0       0   
Albania                   0       0       0       0       0       0       0   
Algeria                   0       0       0       0       0       0       0   
Andorra                   0       0       0       0       0       0       0   
Antigua and Barbuda       0       0       0       0       0       0       0   

                     2/10/20  2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  \
Country/Region                                                              
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   

                     2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
Country/Region                                                              
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   

                     2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  \
Country/Region                                                              
Afghanistan                0        0        1        1        1        1   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        1        1        1   
Andorra                    0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   

                     2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  \
Country/Region                                                                  
Afghanistan                1        1       1       1       1       1       1   
Albania                    0        0       0       0       0       0       0   
Algeria                    1        1       1       3       5      12      12   
Andorra                    0        0       0       1       1       1       1   
Antigua and Barbuda        0        0       0       0       0       0       0   

                     3/6/20  3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  \
Country/Region                                                          
Afghanistan               1       1       4       4        5       

In [41]:
dropped_df_confirmed_total.tail()

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  \
Country/Region                                                                  
Uruguay               0        0        0        0        0        0        0   
Uzbekistan            0        0        0        0        0        0        0   
Venezuela             0        0        0        0        0        0        0   
Vietnam               0        2        2        2        2        2        2   
Zambia                0        0        0        0        0        0        0   

                1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  \
Country/Region                                                              
Uruguay               0        0        0       0       0       0       0   
Uzbekistan            0        0        0       0       0       0       0   
Venezuela             0        0        0       0       0       0       0   
Vietnam               2        2        2       6       6       8       8   
Zambia                0        0        0       0       0       0       0   

                2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  \
Country/Region                                                             
Uruguay              0       0       0       0       0        0        0   
Uzbekistan           0       0       0       0       0        0        0   
Venezuela            0       0       0       0       0        0        0   
Vietnam              8      10      10      13      13       14       15   
Zambia               0       0       0       0       0        0        0   

                2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
Country/Region                                                                  
Uruguay               0        0        0        0        0        0        0   
Uzbekistan            0        0        0        0        0        0        0   
Venezuela             0        0        0        0        0        0        0   
Vietnam              15       16       16       16       16       16       16   
Zambia                0        0        0        0        0        0        0   

                2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  \
Country/Region                                                                  
Uruguay               0        0        0        0        0        0        0   
Uzbekistan            0        0        0        0        0        0        0   
Venezuela             0        0        0        0        0        0        0   
Vietnam              16       16       16       16       16       16       16   
Zambia                0        0        0        0        0        0        0   

                2/26/20  2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  \
Country/Region                                                               
Uruguay               0        0        0        0       0       0       0   
Uzbekistan            0        0        0        0       0       0       0   
Venezuela             0        0        0        0       0       0       0   
Vietnam              16       16       16       16      16      16      16   
Zambia                0        0        0        0       0       0       0   

                3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  3/10/20  \
Country/Region                                                            
Uruguay              0       0       0       0       0       0        0   
Uzbekistan           0       0       0       0       0       0        0   
Venezuela            0       0       0       0       0       0        0   
Vietnam             16      16      16      18      30      30       31   
Zambia               0       0       0       0       0       0        0   

                3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
Country/Region                                                                  
Urug

In [42]:
dropped_df_confirmed_total.describe()


1/22/20     1/23/20     1/24/20      1/25/20      1/26/20  \
count  152.000000  152.000000  152.000000   152.000000   152.000000   
mean     3.651316    4.296053    6.190789     9.434211    13.934211   
std     44.445705   52.149916   74.613240   114.029430   168.285183   
min      0.000000    0.000000    0.000000     0.000000     0.000000   
25%      0.000000    0.000000    0.000000     0.000000     0.000000   
50%      0.000000    0.000000    0.000000     0.000000     0.000000   
75%      0.000000    0.000000    0.000000     0.000000     0.000000   
max    548.000000  643.000000  920.000000  1406.000000  2075.000000   

           1/27/20      1/28/20      1/29/20      1/30/20      1/31/20  \
count   152.000000   152.000000   152.000000   152.000000   152.000000   
mean     19.256579    36.697368    40.565789    54.171053    65.309211   
std     233.331685   446.805187   493.681892   660.275838   794.984920   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max    2877.000000  5509.000000  6087.000000  8141.000000  9802.000000   

             2/1/20        2/2/20        2/3/20        2/4/20        2/5/20  \
count    152.000000    152.000000    152.000000    152.000000    152.000000   
mean      79.197368    110.440789    130.796053    157.184211    181.809211   
std      964.413915   1348.791098   1599.094021   1922.796620   2225.577442   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max    11891.000000  16630.000000  19716.000000  23707.000000  27440.000000   

             2/6/20        2/7/20        2/8/20        2/9/20       2/10/20  \
count    152.000000    152.000000    152.000000    152.000000    152.000000   
mean     202.743421    226.256579    244.210526    264.144737    281.328947   
std     2480.816020   2766.543890   2985.854101   3230.395185   3435.166336   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max    30587.000000  34110.000000  36814.000000  39829.000000  42354.000000   

            2/11/20       2/12/20       2/13/20       2/14/20       2/15/20  \
count    152.000000    152.000000    152.000000    152.000000    152.000000   
mean     294.750000    297.506579    397.157895    440.032895    454.144737   
std     3599.978317   3630.219168   4857.897230   5382.095187   5548.749148   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max    44386.000000  44759.000000  59895.000000  66358.000000  68413.000000   

            2/16/20       2/17/20       2/18/20       2/19/20       2/20/20  \
count    152.000000    152.000000    152.000000    152.000000    152.000000   
mean     468.578947    481.960526    494.315789    497.625000    501.296053   
std     5719.056339   5874.852515   6018.977649   6052.069635   6089.177766   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.0

In [43]:
dropped_df_confirmed.describe()

1/22/20     1/23/20     1/24/20     1/25/20      1/26/20  \
count  462.000000  462.000000  462.000000  462.000000   462.000000   
mean     1.201299    1.413420    2.036797    3.103896     4.584416   
std     20.708297   20.792959   25.900208   36.062942    50.412877   
min      0.000000    0.000000    0.000000    0.000000     0.000000   
25%      0.000000    0.000000    0.000000    0.000000     0.000000   
50%      0.000000    0.000000    0.000000    0.000000     0.000000   
75%      0.000000    0.000000    0.000000    0.000000     0.000000   
max    444.000000  444.000000  549.000000  761.000000  1058.000000   

           1/27/20      1/28/20      1/29/20      1/30/20      1/31/20  \
count   462.000000   462.000000   462.000000   462.000000   462.000000   
mean      6.335498    12.073593    13.346320    17.822511    21.487013   
std      67.934111   166.507245   167.590443   230.917736   273.793450   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max    1423.000000  3554.000000  3554.000000  4903.000000  5806.000000   

            2/1/20        2/2/20        2/3/20        2/4/20        2/5/20  \
count   462.000000    462.000000    462.000000    462.000000    462.000000   
mean     26.056277     36.335498     43.032468     51.714286     59.816017   
std     336.896685    523.345352    632.611813    779.603278    918.652955   
min       0.000000      0.000000      0.000000      0.000000      0.000000   
25%       0.000000      0.000000      0.000000      0.000000      0.000000   
50%       0.000000      0.000000      0.000000      0.000000      0.000000   
75%       0.000000      0.000000      0.000000      0.000000      0.000000   
max    7153.000000  11177.000000  13522.000000  16678.000000  19665.000000   

             2/6/20        2/7/20        2/8/20        2/9/20       2/10/20  \
count    462.000000    462.000000    462.000000    462.000000    462.000000   
mean      66.703463     74.439394     80.346320     86.904762     92.558442   
std     1032.660278   1164.900128   1264.888657   1382.595175   1480.138612   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max    22112.000000  24953.000000  27100.000000  29631.000000  31728.000000   

            2/11/20       2/12/20       2/13/20       2/14/20       2/15/20  \
count    462.000000    462.000000    462.000000    462.000000    462.000000   
mean      96.974026     97.880952    130.666667    144.772727    149.415584   
std     1556.359217   1556.658334   2245.535032   2533.645344   2619.375809   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max    33366.000000  33366.000000  48206.000000  54406.000000  56249.000000   

            2/16/20       2/17/20       2/18/20       2/19/20       2/20/20  \
count    462.000000    462.000000    462.000000    462.000000    462.000000   
mean     154.164502    158.567100    162.632035    163.720779    164.928571   
std     2709.266560   2793.286362   2872.003896   2888.262233   2907.369467   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.00000

In [44]:
dropped_df_confirmed_total.describe(include=['object', 'bool'])

ValueError: No objects to concatenate

In [45]:
df_deaths.head()

Province/State Country/Region      Lat      Long  1/22/20  1/23/20  1/24/20  \
0            NaN       Thailand  15.0000  101.0000        0        0        0   
1            NaN          Japan  36.0000  138.0000        0        0        0   
2            NaN      Singapore   1.2833  103.8333        0        0        0   
3            NaN          Nepal  28.1667   84.2500        0        0        0   
4            NaN       Malaysia   2.5000  112.5000        0        0        0   

   1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  \
0        0        0        0        0        0        0        0       0   
1        0        0        0        0        0        0        0       0   
2        0        0        0        0        0        0        0       0   
3        0        0        0        0        0        0        0       0   
4        0        0        0        0        0        0        0       0   

   2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  \
0       0       0       0       0       0       0       0       0        0   
1       0       0       0       0       0       0       0       0        0   
2       0       0       0       0       0       0       0       0        0   
3       0       0       0       0       0       0       0       0        0   
4       0       0       0       0       0       0       0       0        0   

   2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        1        1        1        1        1        1   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  \
0        0        0        0        0        0        0        0        0   
1        1        1        1        1        1        1        1        2   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  \
0        0        0        0       1       1       1       1       1       1   
1        4        4        5       6       6       6       6       6       6   
2        0        0        0       0       0       0       0       0       0   
3        0        0        0       0       0       0       0       0       0   
4        0        0        0       0       0       0       0       0       0   

   3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  \
0       1       1       1        1        1        1        1        1   
1       6       6      10       10       15       16       19       22   
2       0       0       0        0        0        0        0        0   
3       0       0       0        0        0        0        0        0   
4       0       0       0        0        0        0        0        0   

   3/15/20  3/16/20  3/17/20  3/18/20  
0        1        1        1        1  
1       22       27       29       29  
2        0        0        0        0  
3        0        0        0        0  
4        0        0        2        2

In [46]:
dropped_df_deaths.describe(include=['object', 'bool'])

NameError: name 'dropped_df_deaths' is not defined

In [47]:
# drop columns that we don't need them
# we focus on data by Country/Region only
dropped_df_deaths = df_deaths.drop(['Province/State','Lat', 'Long'], axis=1)
dropped_df_deaths.head()

Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0       Thailand        0        0        0        0        0        0   
1          Japan        0        0        0        0        0        0   
2      Singapore        0        0        0        0        0        0   
3          Nepal        0        0        0        0        0        0   
4       Malaysia        0        0        0        0        0        0   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   
2        0        0        0        0       0       0       0       0       0   
3        0        0        0        0       0       0       0       0       0   
4        0        0        0        0       0       0       0       0       0   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        1   
2       0       0       0       0        0        0        0        0   
3       0       0       0       0        0        0        0        0   
4       0       0       0       0        0        0        0        0   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0        0        0        0        0        0        0        0        0   
1        1        1        1        1        1        1        1        1   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0        0        0        0        0        0        0        0        0   
1        1        1        1        1        2        4        4        5   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  \
0       1       1       1       1       1       1       1       1       1   
1       6       6       6       6       6       6       6       6      10   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
0        1        1        1        1        1        1        1        1   
1       10       15       16       19       22       22       27       29   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        2   

   3/18/20  
0        1  
1       29  
2        0  
3        0  
4        2

In [48]:
dropped_df_deaths_total = dropped_df_deaths.groupby(['Country/Region']).sum()
dropped_df_deaths_total.head()

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
Country/Region                                                              
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   

                     1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  \
Country/Region                                                            
Afghanistan                0        0        0        0       0       0   
Albania                    0        0        0        0       0       0   
Algeria                    0        0        0        0       0       0   
Andorra                    0        0        0        0       0       0   
Antigua and Barbuda        0        0        0        0       0       0   

                     2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  \
Country/Region                                                                
Afghanistan               0       0       0       0       0       0       0   
Albania                   0       0       0       0       0       0       0   
Algeria                   0       0       0       0       0       0       0   
Andorra                   0       0       0       0       0       0       0   
Antigua and Barbuda       0       0       0       0       0       0       0   

                     2/10/20  2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  \
Country/Region                                                              
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   

                     2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
Country/Region                                                              
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   

                     2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  \
Country/Region                                                              
Afghanistan                0        0        0        0        0        0   
Albania                    0        0        0        0        0        0   
Algeria                    0        0        0        0        0        0   
Andorra                    0        0        0        0        0        0   
Antigua and Barbuda        0        0        0        0        0        0   

                     2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  \
Country/Region                                                                  
Afghanistan                0        0       0       0       0       0       0   
Albania                    0        0       0       0       0       0       0   
Algeria                    0        0       0       0       0       0       0   
Andorra                    0        0       0       0       0       0       0   
Antigua and Barbuda        0        0       0       0       0       0       0   

                     3/6/20  3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  \
Country/Region                                                          
Afghanistan               0       0       0       0        0       

In [49]:
dropped_df_deaths_total.tail()

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  \
Country/Region                                                                  
Uruguay               0        0        0        0        0        0        0   
Uzbekistan            0        0        0        0        0        0        0   
Venezuela             0        0        0        0        0        0        0   
Vietnam               0        0        0        0        0        0        0   
Zambia                0        0        0        0        0        0        0   

                1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  \
Country/Region                                                              
Uruguay               0        0        0       0       0       0       0   
Uzbekistan            0        0        0       0       0       0       0   
Venezuela             0        0        0       0       0       0       0   
Vietnam               0        0        0       0       0       0       0   
Zambia                0        0        0       0       0       0       0   

                2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  \
Country/Region                                                             
Uruguay              0       0       0       0       0        0        0   
Uzbekistan           0       0       0       0       0        0        0   
Venezuela            0       0       0       0       0        0        0   
Vietnam              0       0       0       0       0        0        0   
Zambia               0       0       0       0       0        0        0   

                2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
Country/Region                                                                  
Uruguay               0        0        0        0        0        0        0   
Uzbekistan            0        0        0        0        0        0        0   
Venezuela             0        0        0        0        0        0        0   
Vietnam               0        0        0        0        0        0        0   
Zambia                0        0        0        0        0        0        0   

                2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  \
Country/Region                                                                  
Uruguay               0        0        0        0        0        0        0   
Uzbekistan            0        0        0        0        0        0        0   
Venezuela             0        0        0        0        0        0        0   
Vietnam               0        0        0        0        0        0        0   
Zambia                0        0        0        0        0        0        0   

                2/26/20  2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  \
Country/Region                                                               
Uruguay               0        0        0        0       0       0       0   
Uzbekistan            0        0        0        0       0       0       0   
Venezuela             0        0        0        0       0       0       0   
Vietnam               0        0        0        0       0       0       0   
Zambia                0        0        0        0       0       0       0   

                3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  3/10/20  \
Country/Region                                                            
Uruguay              0       0       0       0       0       0        0   
Uzbekistan           0       0       0       0       0       0        0   
Venezuela            0       0       0       0       0       0        0   
Vietnam              0       0       0       0       0       0        0   
Zambia               0       0       0       0       0       0        0   

                3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
Country/Region                                                                  
Urug

In [50]:
dropped_df_confirmed_total.describe()


1/22/20     1/23/20     1/24/20      1/25/20      1/26/20  \
count  152.000000  152.000000  152.000000   152.000000   152.000000   
mean     3.651316    4.296053    6.190789     9.434211    13.934211   
std     44.445705   52.149916   74.613240   114.029430   168.285183   
min      0.000000    0.000000    0.000000     0.000000     0.000000   
25%      0.000000    0.000000    0.000000     0.000000     0.000000   
50%      0.000000    0.000000    0.000000     0.000000     0.000000   
75%      0.000000    0.000000    0.000000     0.000000     0.000000   
max    548.000000  643.000000  920.000000  1406.000000  2075.000000   

           1/27/20      1/28/20      1/29/20      1/30/20      1/31/20  \
count   152.000000   152.000000   152.000000   152.000000   152.000000   
mean     19.256579    36.697368    40.565789    54.171053    65.309211   
std     233.331685   446.805187   493.681892   660.275838   794.984920   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max    2877.000000  5509.000000  6087.000000  8141.000000  9802.000000   

             2/1/20        2/2/20        2/3/20        2/4/20        2/5/20  \
count    152.000000    152.000000    152.000000    152.000000    152.000000   
mean      79.197368    110.440789    130.796053    157.184211    181.809211   
std      964.413915   1348.791098   1599.094021   1922.796620   2225.577442   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max    11891.000000  16630.000000  19716.000000  23707.000000  27440.000000   

             2/6/20        2/7/20        2/8/20        2/9/20       2/10/20  \
count    152.000000    152.000000    152.000000    152.000000    152.000000   
mean     202.743421    226.256579    244.210526    264.144737    281.328947   
std     2480.816020   2766.543890   2985.854101   3230.395185   3435.166336   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max    30587.000000  34110.000000  36814.000000  39829.000000  42354.000000   

            2/11/20       2/12/20       2/13/20       2/14/20       2/15/20  \
count    152.000000    152.000000    152.000000    152.000000    152.000000   
mean     294.750000    297.506579    397.157895    440.032895    454.144737   
std     3599.978317   3630.219168   4857.897230   5382.095187   5548.749148   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max    44386.000000  44759.000000  59895.000000  66358.000000  68413.000000   

            2/16/20       2/17/20       2/18/20       2/19/20       2/20/20  \
count    152.000000    152.000000    152.000000    152.000000    152.000000   
mean     468.578947    481.960526    494.315789    497.625000    501.296053   
std     5719.056339   5874.852515   6018.977649   6052.069635   6089.177766   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.0

In [51]:
dropped_df_deaths.describe()

1/22/20     1/23/20     1/24/20     1/25/20     1/26/20     1/27/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean     0.036797    0.038961    0.056277    0.090909    0.121212    0.177489   
std      0.790912    0.792178    1.118312    1.861927    2.420633    3.537051   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max     17.000000   17.000000   24.000000   40.000000   52.000000   76.000000   

          1/28/20     1/29/20     1/30/20     1/31/20      2/1/20      2/2/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean     0.283550    0.287879    0.370130    0.461039    0.560606    0.783550   
std      5.816025    5.816558    7.537859    9.491496   11.584810   16.283441   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max    125.000000  125.000000  162.000000  204.000000  249.000000  350.000000   

           2/3/20      2/4/20      2/5/20      2/6/20      2/7/20      2/8/20  \
count  462.000000  462.000000  462.000000  462.000000  462.000000  462.000000   
mean     0.922078    1.064935    1.220779    1.372294    1.556277    1.744589   
std     19.260808   22.284692   25.541149   28.751307   32.519577   36.288129   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max    414.000000  479.000000  549.000000  618.000000  699.000000  780.000000   

           2/9/20     2/10/20      2/11/20      2/12/20      2/13/20  \
count  462.000000  462.000000   462.000000   462.000000   462.000000   
mean     1.961039    2.192641     2.409091     2.419913     2.967532   
std     40.521902   45.313702    49.687127    49.687233    60.945666   
min      0.000000    0.000000     0.000000     0.000000     0.000000   
25%      0.000000    0.000000     0.000000     0.000000     0.000000   
50%      0.000000    0.000000     0.000000     0.000000     0.000000   
75%      0.000000    0.000000     0.000000     0.000000     0.000000   
max    871.000000  974.000000  1068.000000  1068.000000  1310.000000   

           2/14/20      2/15/20      2/16/20      2/17/20      2/18/20  \
count   462.000000   462.000000   462.000000   462.000000   462.000000   
mean      3.296537     3.606061     3.831169     4.043290     4.344156   
std      67.784979    74.251731    78.903561    83.230766    89.372290   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max    1457.000000  1596.000000  1696.000000  1789.000000  1921.000000   

           2/19/20      2/20/20      2/21/20      2/22/20      2/23/20  \
count   462.000000   462.000000   462.000000   462.000000   462.000000   
mean      4.593074     4.863636     4.872294     5.320346     5.344156   
std      94.396238    99.745451    99.745201   109.142175   109.142013   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000   

In [52]:
df_recovered.head()

Province/State Country/Region      Lat      Long  1/22/20  1/23/20  1/24/20  \
0            NaN       Thailand  15.0000  101.0000        0        0        0   
1            NaN          Japan  36.0000  138.0000        0        0        0   
2            NaN      Singapore   1.2833  103.8333        0        0        0   
3            NaN          Nepal  28.1667   84.2500        0        0        0   
4            NaN       Malaysia   2.5000  112.5000        0        0        0   

   1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  \
0        0        2        2        5        5        5        5       5   
1        0        1        1        1        1        1        1       1   
2        0        0        0        0        0        0        0       0   
3        0        0        0        0        0        0        0       0   
4        0        0        0        0        0        0        0       0   

   2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  \
0       5       5       5       5       5       5      10      10       10   
1       1       1       1       1       1       1       1       1        4   
2       0       0       0       0       0       0       2       2        2   
3       0       0       0       0       0       0       0       0        0   
4       0       0       0       0       0       1       1       1        1   

   2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
0       10       10       12       12       12       14       15       15   
1        9        9        9        9       12       12       12       13   
2        9       15       15       17       18       18       24       29   
3        0        1        1        1        1        1        1        1   
4        3        3        3        3        7        7        7       13   

   2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  \
0       15       15       17       17       21       21       22       22   
1       18       18       22       22       22       22       22       22   
2       34       34       37       37       51       51       53       62   
3        1        1        1        1        1        1        1        1   
4       15       15       15       15       15       18       18       18   

   2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  \
0       22       28       28      28      31      31      31      31      31   
1       22       22       32      32      32      43      43      43      46   
2       62       62       72      72      78      78      78      78      78   
3        1        1        1       1       1       1       1       1       1   
4       18       18       18      18      18      22      22      22      22   

   3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  \
0      31      31      31       33       34       34       35       35   
1      76      76      76      101      118      118      118      118   
2      78      78      78       78       96       96       97      105   
3       1       1       1        1        1        1        1        1   
4      23      24      24       24       26       26       26       35   

   3/15/20  3/16/20  3/17/20  3/18/20  
0       35       35       41       42  
1      118      144      144      144  
2      105      109      114      114  
3        1        1        1        1  
4       42       42       49       60

In [ ]:
# drop columns that we don't need them
# we focus on data by Country/Region only
dropped_df_recovered = df_recovered.drop(['Province/State','Lat', 'Long'], axis=1)
dropped_df_recovered.head()

In [ ]:
dropped_df_recovered_total = dropped_df_recovered.groupby(['Country/Region']).sum()
dropped_df_recovered_total.head()

In [ ]:
dropped_df_recovered_total.tail()

In [ ]:
dropped_df_recovered_total.describe()


In [ ]:
dropped_df_recovered.describe()

### STEP 5: MELTING DATA & MERGING DATA

In [ ]:
def melt_data(df):
    """
    melt data of one zip code from wide format to long format
    """
    
    melted = pd.melt(df, id_vars=['Country/Region'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True)
    #melted = melted.dropna(subset=['value'])
    
    return melted.groupby('time').aggregate({'value':'mean'})

In [ ]:
def melt_df(df):
    """
    Loop through all zipcodes to melt data of each zipcode.
    Then, merge all melted data back together
    """
    
    merged = []
    for country in df['Country/Region']:
        melted = melt_data(df.loc[df['Country/Region'] == country])
        row = df.loc[df['Country/Region'] == country].iloc[:,:1]
        rows = pd.concat([row]*len(melted), ignore_index=True)
        merge = pd.concat([rows, melted.reset_index()], axis= 1)
        merged.append(merge)
    melted_df = pd.concat(merged)
    return melted_df

In [ ]:
dropped_df_confirmed_total.reset_index(inplace=True)

In [ ]:
dropped_df_confirmed_total.head()

In [ ]:
df = melt_df(dropped_df_confirmed_total)

In [ ]:
df.head()

In [ ]:
# check any columns has na/nan value, if there is missing data
#### forward fill missing value
#df['time'] = df['time'].ffill()
df.isnull().any()

We see there are some missing data in time and value column. We need to fill these data.

In [ ]:
# make it as time series
df.set_index('time', inplace=True)

In [ ]:
# look at the head again
df.head()

In [ ]:
# check any columns has na/nan value, if there is missing data
df.isnull().any()

We see value column has missing data
It’s often useful to be able to fill missing data with realistic values such as the average of a time period. However, if we are working with a time series problem and we really want our data to be realistic, we wouldn't want to do a backfill of our data since that’s like looking into the future and getting information you would never have at that time period. We are going to do forward fill our data as follow.

In [ ]:
# rename columns
df.rename(columns={'Country/Region': 'Country'}, inplace=True)
df.head()

# Step 2: EDA and Visualization

In [ ]:
# countries affected
countries = df['Country'].unique()
len(countries)

In [ ]:
plt.figure(figsize=(30,18))

for c in countries:
    df[df['Country']==c]['value'].plot(label=c)

plt.legend(loc='upper center', bbox_to_anchor=(0.5, -0.05), fancybox=True, shadow=True, ncol=10)
plt.show()